### PR TITLE
Standardize sending message :

### PR DESCRIPTION
**Pull request by Zowac**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Fixed issues | comma-separated list of issues fixed by the PR, if any

### CODE REVIEW

File: src/main/core/API/Finder/Log/MessageLogFinder.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    public function getClass()
+    {
+        return MessageLog::class;
+    }
+
+    public function configureQueryBuilder(QueryBuilder $qb, array $searches = [], array $sortBy = null, array $options = ['count' => false, 'page' => 0, 'limit' => -1])
+    {
+        foreach ($searches as $filterName => $filterValue) {
+            switch ($filterName) {
+                case 'workspace':
```
Comment (Elorfin) [inline src/main/core/API/Finder/Log/MessageLogFinder.php:29]: This prop does not exist in your entity.

File: src/plugin/cursus/Manager/EventManager.php [hunk 2]
```diff
@@ -52,14 +57,16 @@ public function __construct(
         ObjectManager $om,
         UrlGeneratorInterface $router,
         TemplateManager $templateManager,
-        TokenStorageInterface $tokenStorage
+        TokenStorageInterface $tokenStorage,
+        StrictDispatcher $dispatcher
```
Comment (Elorfin) [inline src/plugin/cursus/Manager/EventManager.php:61]: MailManager injection can be removed.

File: src/plugin/cursus/Manager/SessionManager.php [hunk 3]
```diff
@@ -82,7 +87,8 @@ public function __construct(
         TemplateManager $templateManager,
         TokenStorageInterface $tokenStorage,
         WorkspaceManager $workspaceManager,
-        EventManager $sessionEventManager
+        EventManager $sessionEventManager,
+        StrictDispatcher $dispatcher
```
Comment (Elorfin) [inline src/plugin/cursus/Manager/SessionManager.php:91]: MailManager injection can be removed.

File: src/plugin/message/Manager/MessageManager.php [hunk 4]
```diff
@@ -178,22 +178,25 @@ public function send(Message $message, $setAsSent = true, $sendMail = true)
         return $message;
     }
 
-    public function sendMessageToAbstractRoleSubject(
-        AbstractRoleSubject $subject,
+    public function sendMessage(
         $content,
         $object,
+        ?AbstractRoleSubject $subject = null,
+        $users = [],
```
Comment (Elorfin) [inline src/plugin/message/Manager/MessageManager.php:185]: You should only keep an array of AbstractRoleSubject and name it `receivers`.

File: src/plugin/message/Subscriber/MessageSubscriber.php [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        } else {
+            $logEntry = new MessageLog();
+            $logEntry->setDetails($event->getMessage($this->translator, $sender, $event->getReceiver()));
+            $logEntry->setEvent($eventName);
+            $logEntry->setReceiver($event->getReceiver());
+            $logEntry->setSender($sender);
+
+            $this->em->persist($logEntry);
+
+            $this->em->persist($logEntry);
```
Comment (Elorfin) [inline src/plugin/message/Subscriber/MessageSubscriber.php:81]: Duplicated persist.

File: src/plugin/message/Subscriber/MessageSubscriber.php [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+            foreach ($event->getUsers() as $user) {
+                $logEntry = new MessageLog();
+                $logEntry->setDetails($event->getMessage($this->translator, $sender, $user));
+                $logEntry->setEvent($eventName);
+                $logEntry->setReceiver($user);
+                $logEntry->setSender($sender);
+
+                $this->em->persist($logEntry);
+            }
+        } else {
```
Comment (Elorfin) [inline src/plugin/message/Subscriber/MessageSubscriber.php:72]: Else will no longer be required after merging props.

**Review by Elorfin — CHANGES_REQUESTED** [review]
Good work. Just some little changes are requested.

File: src/main/core/API/Finder/Log/MessageLogFinder.php (hunk 1 above)
Comment (Zowac) [reply src/main/core/API/Finder/Log/MessageLogFinder.php]: Change done, thx :)

File: src/plugin/cursus/Manager/EventManager.php (hunk 2 above)
Comment (Zowac) [reply src/plugin/cursus/Manager/EventManager.php:57]: Done :)

File: src/plugin/message/Manager/MessageManager.php (hunk 4 above)
Comment (Zowac) [reply src/plugin/message/Manager/MessageManager.php]: Done :) 
thx

File: src/plugin/message/Manager/MessageManager.php [hunk 7]
```diff
@@ -178,26 +178,17 @@ public function send(Message $message, $setAsSent = true, $sendMail = true)
...
-        }
-
-        if ($subject instanceof Group) {
-            foreach ($subject->getUsers() as $user) {
-                $users[] = $user;
-            }
-        }
-
-        $message = $this->create($content, $object, $users, $sender);
+        $message = $this->create($content, $object, $receivers, $sender);
```
Comment (Elorfin) [inline src/plugin/message/Manager/MessageManager.php:191]: Create method expects an array of `User`, but receive an array of `AbstractRoleSubject` which can also be `Group`. You must keep the previous logic on the `$subject` var and apply it to all the elements of the `$receivers` array.

File: src/main/core/API/Finder/Log/MessageLogFinder.php [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    public function configureQueryBuilder(QueryBuilder $qb, array $searches = [], array $sortBy = null, array $options = ['count' => false, 'page' => 0, 'limit' => -1])
+    {
+        foreach ($searches as $filterName => $filterValue) {
+            switch ($filterName) {
+                case 'sender':
+                    $qb->andWhere('UPPER(obj.username) LIKE :name OR UPPER(CONCAT(obj.firstName, \' \', obj.lastName)) LIKE :name');
+                    $qb->setParameter('sender', '%'.strtoupper($filterValue).'%');
+                    break;
```
Comment (Elorfin) [inline src/main/core/API/Finder/Log/MessageLogFinder.php:32]: Search must be done on `uuid`.

File: src/plugin/message/Subscriber/MessageSubscriber.php [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+            $event->getContent(),
+            $event->getObject(),
+            $event->getReceivers(),
+            $event->getSender(),
+            $event->getWithMail()
+        );
+
+        $sender = $event->getSender() ?? $this->security->getUser();
+
+        foreach ($event->getReceivers() as $user) {
```
Comment (Elorfin) [inline src/plugin/message/Subscriber/MessageSubscriber.php:61]: I think you may have `Group` in this array. so `$logEntry->setReceiver($user);` will not work.

File: src/plugin/planned-notification/Manager/PlannedNotificationManager.php [hunk 10]
```diff
@@ -157,7 +165,15 @@ public function generateScheduledTasks(
...
     {
         foreach ($messages as $message) {
-            $this->mailManager->send($message->getTitle(), $message->getContent(), $users);
+            $this->dispatcher->dispatch(
+                MessageEvents::MESSAGE_SENDING,
+                SendMessageEvent::class,
+                [
+                    $message->getContent(),
+                    $message->getTitle(),
+                    [$users],
```
Comment (Elorfin) [inline src/plugin/planned-notification/Manager/PlannedNotificationManager.php:174]: Array of array. IMO the wrapping array must be removed.

**Review by Elorfin — CHANGES_REQUESTED** [review]

File: src/plugin/planned-notification/Manager/PlannedNotificationManager.php (hunk 10 above)
Comment (Zowac) [reply src/plugin/planned-notification/Manager/PlannedNotificationManager.php:174]: Oops my bad, fixed. Thx :)

File: src/plugin/message/Subscriber/MessageSubscriber.php (hunk 9 above)
Comment (Zowac) [reply src/plugin/message/Subscriber/MessageSubscriber.php:61]: Yes thx, I fixed It

File: src/main/core/API/Finder/Log/MessageLogFinder.php (hunk 8 above)
Comment (Zowac) [reply src/main/core/API/Finder/Log/MessageLogFinder.php:32]: Done

File: src/plugin/message/Manager/MessageManager.php (hunk 7 above)
Comment (Zowac) [reply src/plugin/message/Manager/MessageManager.php:191]: I reverted my change about the previous logic, thx

File: src/main/core/API/Finder/Log/MessageLogFinder.php (hunk 8 above)
Comment (Elorfin) [reply src/main/core/API/Finder/Log/MessageLogFinder.php:32]: Sorry, I was saying the user uuid (previous search was done on user names).

**Review by Elorfin — CHANGES_REQUESTED** [review]

**Review by Elorfin — APPROVED** [review]

**Pull request merged**